openff/toolkit/topology/molecule.py

```diff
@@ -1255,13 +1255,23 @@ def __hash__(self):
         return hash(self.to_smiles())
 
     def ordered_connection_table_hash(self) -> int:
-        """Compute an ordered hash of the atoms and bonds in the molecule"""
+        """
+        Compute an ordered hash of the atoms and bonds in the molecule.
+        This hash method is intended for comparison of Molecule objects at
+        runtime, and hashes from one version of the software should not be
+        compared with hashes generated using different versions.
+        """
         if self._ordered_connection_table_hash is not None:
             return self._ordered_connection_table_hash
 
+        # Pre-assign molecule atom indices in O(N) time to avoid use of List.index to get index of each one
+        # in O(N^2) time
+        for index, atom in enumerate(self.atoms):
+            atom._molecule_atom_index = index
+
```

Comment:
This could be moved to `_add_atom()`

Reply:
That's a complicated enough change (will become wrong if we ever allow more than just appending atoms, then this would need to be handled as an invalidatable cached property) that I'm gonna leave this as-is until a significant runtime benefit is demonstrated.